### ISSUE TITLE
UnboundLocalError: local variable 'url' referenced before assignment

### ISSUE THREAD

**Issue by martinpitt** [teemtee/tmt#1657]
I just ran my standard tmt command for Fedora 37, and it crashes now:
```
❱❱❱ tmt --debug run --all provision --how virtual --image fedora-37
/var/tmp/tmt/run-004

/plans/all
    Report step always force mode enabled.
    discover
    workdir: /var/tmp/tmt/run-004/plans/all/discover
        how: fmf
        directory: /var/home/martin/upstream/python-dbusmock
        Copy '/var/home/martin/upstream/python-dbusmock' to '/var/tmp/tmt/run-004/plans/all/discover/default-0/tests'.
        tests: /tests/smoke
        Check metadata tree in '/var/tmp/tmt/run-004/plans/all/discover/default-0/tests/'.
        summary: 1 test selected
    provision
    workdir: /var/tmp/tmt/run-004/plans/all/provision
        how: virtual
        user: root
        image: fedora-37
        memory: 2048 MB
        disk: 10 GB
        arch: x86_64
    finish
    workdir: /var/tmp/tmt/run-004/plans/all/finish
        summary: 0 tasks completed
Traceback (most recent call last):
  File "/usr/bin/tmt", line 8, in <module>
    tmt.cli.main()
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1691, in invoke
    return _process_result(rv)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 1628, in _process_result
    value = ctx.invoke(self._result_callback, value, **ctx.params)
  File "/usr/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3.10/site-packages/tmt/cli.py", line 406, in finito
    click_context.obj.run.go()
  File "/usr/lib/python3.10/site-packages/tmt/base.py", line 2057, in go
    plan.go()
  File "/usr/lib/python3.10/site-packages/tmt/base.py", line 1210, in go
    step.go()
  File "/usr/lib/python3.10/site-packages/tmt/steps/provision/__init__.py", line 226, in go
    phase.go()
  File "/usr/lib/python3.10/site-packages/tmt/steps/provision/testcloud.py", line 313, in go
    self._guest.start()
  File "/usr/lib/python3.10/site-packages/tmt/steps/provision/testcloud.py", line 536, in start
    self.image_url = self._guess_image_url(self.image_url)
  File "/usr/lib/python3.10/site-packages/tmt/steps/provision/testcloud.py", line 434, in _guess_image_url
    url = testcloud.util.get_fedora_image_url(
  File "/usr/lib/python3.10/site-packages/testcloud/util.py", line 109, in get_fedora_image_url
    return url
UnboundLocalError: local variable 'url' referenced before assignment
```

This has worked before -- not sure if it's something that changed in tmt, testcloud, the Fedora mirror/metadata, etc. It does work with `fedora-rawhide` and also with `fedora-36`, just `fedora-37` seems to be "special" at the moment.

This could of course be an actually broken F37 metadata, but I suppose the error message should be a bit more friendly?

tmt-1.18.0-1.fc36.noarch
python3-testcloud-0.8.1-1.fc36.noarch

**Comment by martinpitt** [comment]
I re-tested this in a Fedora 37 toolbox (original report was in Fedora 36). F37 has an *older* version of tmt (!), 1.17.0-1. And a sligthly newer testcloud: python3-testcloud-0.8.1-2.fc37.noarch

The crash happens there as well. It looks a little more detailed, as Python 3.11 has better contexts:

```
Traceback (most recent call last):
  File "/usr/bin/tmt", line 8, in <module>
    tmt.cli.main()
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1689, in invoke
    return _process_result(rv)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1626, in _process_result
    value = ctx.invoke(self._result_callback, value, **ctx.params)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/tmt/cli.py", line 406, in finito
    click_context.obj.run.go()
  File "/usr/lib/python3.11/site-packages/tmt/base.py", line 2293, in go
    plan.go()
  File "/usr/lib/python3.11/site-packages/tmt/base.py", line 1454, in go
    step.go()
  File "/usr/lib/python3.11/site-packages/tmt/steps/provision/__init__.py", line 224, in go
    phase.go()
  File "/usr/lib/python3.11/site-packages/tmt/steps/provision/testcloud.py", line 309, in go
    self._guest.start()
  File "/usr/lib/python3.11/site-packages/tmt/steps/provision/testcloud.py", line 531, in start
    self.image_url = self._guess_image_url(self.image_url)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/tmt/steps/provision/testcloud.py", line 430, in _guess_image_url
    url = testcloud.util.get_fedora_image_url(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/testcloud/util.py", line 109, in get_fedora_image_url
    return url
           ^^^
UnboundLocalError: cannot access local variable 'url' where it is not associated with a value
```

**Comment by idorax** [comment]
From the source code, looks we should catch the exception(s) from `testcloud.util.get_XXX_image_url()`.
* https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L430
  ```python
   246	    def _guess_image_url(self, name: str) -> str:
   247	        """ Guess image url for given name """
   ...
   266	        assert testcloud is not None
   267	        # Plain name match means we want the latest release
   268	        if name == 'fedora':
   269	            url = testcloud.util.get_fedora_image_url("latest", self.arch)
   270	        elif name == 'fedora-coreos':
   271	            url = testcloud.util.get_fedora_image_url("stable", self.arch)
   272	        elif name == 'centos':
   273	            url = testcloud.util.get_centos_image_url("latest", stream=False, arch=self.arch)
   ...
  ```

And `testcloud.util.get_XXX_image_url()` looks fragile.
* https://pagure.io/testcloud/blob/master/f/testcloud/util.py#_68
  ```python
    51	def get_fedora_image_url(version, arch="x86_64"):
    ..
    60	    primary_arches = ["x86_64", "aarch64"]
    61	    # get coreos url
    62	    if version in config_data.STREAM_LIST:
    63	        try:
    64	            result = requests.get("https://builds.coreos.fedoraproject.org/streams/%s.json"%version).json()
    65	        except (ConnectionError, IndexError, requests.exceptions.JSONDecodeError):
    66	              log.error("Failed to fetch the image.")
    67	              return None
    68	        url = result['architectures'][arch]['artifacts']['qemu']['formats']['qcow2.xz']['disk']['location']
    69	
    70	        return url
  ```

Hi @psss, should we file an issue to `testcloud`?

**Comment by frantisekz** [comment]
This should be fixed by https://pagure.io/testcloud/c/60ec4fda57bcffd0cb84de201b2e28de187e1b88?branch=master 

The code is far from bulletproof, but it at least wouldn't outright crash.

**Comment by idorax** [comment]
@frantisekz, thanks for enhancing `testcloud.util`. 
BTW, do you think it is worthwhile catching exception(s) from tmt code side when invoking `testcloud.util.get_XXX_image_url()`? 
If not, I'd like to close this issue once your patch is merged into the master branch of testcloud.

**Comment by ZelenyMartin** [comment]
Sorry for the nois
